provider/docker/docker.go

```diff
@@ -218,16 +219,22 @@ func (p *Provider) Provide(configurationChan chan<- types.ConfigMessage, pool *s
 					}
 
 					eventsc, errc := dockerClient.Events(ctx, options)
-					for event := range eventsc {
-						if event.Action == "start" ||
-							event.Action == "die" ||
-							strings.HasPrefix(event.Action, "health_status") {
-							startStopHandle(event)
+					for {
+						select {
+						case event := <-eventsc:
+							if event.Action == "start" ||
+								event.Action == "die" ||
+								strings.HasPrefix(event.Action, "health_status") {
+								startStopHandle(event)
```

Comment:
Disclaimer: I'm not familiar with the Docker provider.)

Is there a chance that the event channel gets closed (as opposed to receiving a terminating event)? If it does, then the former implementation naturally edited the loop while the new select approach doesn't do so anymore.

Just trying to make sure that we keep terminating when / if needed.

Reply:
There is no chance of the event channel being closed. The channel that is returned by the Events method on the Docker client returns receive-only channels and only ever closes the errors channel.

The errors channel is always pushed into before each of the return paths in the Events method, and always a single error. The only other way that method could be escaped is with a panic which would propagate up to the Traefik provider operation where it would be recovered and retried until success, or until 15 minutes had elapsed (the default exponential backoff limit).

Reply:
I linked it in the PR, but heres the Events method for reference: https://github.com/moby/moby/blob/master/client/events.go#L20-L71

Reply:
Thanks for elaborating. 👍